src/lib/FreeC/LiftedIR/Converter/Type.hs

```diff
@@ -0,0 +1,105 @@
+-- | Implements the IR to lifted IR translation, which applies the monadic lifting
+--   as described by Abel et al. in "Verifying Haskell Programs Using Constructive
+--   Type Theory".
+
+module FreeC.LiftedIR.Converter.Type
+  ( convertFuncArgTypes
+  , convertConArgType
+    -- * Translations
+  , convertType
+  , convertType'
+  )
+where
+
+import qualified FreeC.IR.Syntax               as IR
+import           FreeC.IR.SrcSpan               ( SrcSpan(NoSrcSpan) )
+import qualified FreeC.LiftedIR.Syntax         as LIR
+
+-- | Converts the argument types of a function.
+convertFuncArgTypes
+  :: Maybe Int -- ^ Index of the decreasing argument for recursive functions.
+  -> [IR.Type] -- ^ The argument types.
+  -> [LIR.Type]
+convertFuncArgTypes = maybe convertNonRecFuncArgTypes convertRecFuncArgTypes
+
+-- | Converts the argument types of a non recursive function using @convertType@.
+convertNonRecFuncArgTypes :: [IR.Type] -> [LIR.Type]
+convertNonRecFuncArgTypes = map convertType
+
+-- | Converts the argument types of a recursive function using @convertType@.
+--
+--   The outermost argument at the given index is marked decreasing.
+convertRecFuncArgTypes :: Int -> [IR.Type] -> [LIR.Type]
+convertRecFuncArgTypes decIndex args =
+  let convArgs                      = map convertType args
+      (startArgs, decArg : endArgs) = splitAt (decIndex - 1) convArgs
+  in  startArgs ++ (markOutermostDecreasing decArg : endArgs)
+
+-- | Converts a constructor argument using @convertType@.
+--
+--   All occurrences of the constructed type are marked as structurally smaller.
+convertConArgType :: IR.QName -> IR.Type -> LIR.Type
+convertConArgType ident = markAllDec ident . convertType
+
+-------------------------------------------------------------------------------
+-- Translations                                                              --
+-------------------------------------------------------------------------------
+
+-- | Converts a type from IR to lifted IR by lifting it into the @Free@ monad.
+--
+--   This corresponds to the dagger translation for monotypes as described by
+--   Abel et al.
+--
+--   > τ' = Free τ*
+convertType :: IR.Type -> LIR.Type
+convertType = LIR.FreeTypeCon NoSrcSpan . convertType'
+
+-- | Lifts a type from IR to lifted IR by renaming type variables and
```

Comment:
As far as I can see `convertType'` does not actually rename any type variables or constructors.